### ISSUE TITLE
nop bug

### ISSUE THREAD

**Issue by regehr** [google/souper#304]
We're getting miscompilations from nop synthesis, please take a look @zhengyangl. A reduced test case is attached, the nop optimization in func.cpp (there is only one) is the issue.

[foo.tar.gz](https://github.com/google/souper/files/1752258/foo.tar.gz)

**Comment by zhengyangl** [comment]
Sure, will have a look ASAP.

**Comment by regehr** [comment]
Take a look at the sclang output when SOUPER_DEBUG is set. Souper does 2 constant replacements and then a nop replacement that ends up being a constant due to the earlier optimizations. It all looks OK but maybe something is going wrong in the interaction between optimizations? In general this is something we're going to have to be really careful with, due to Souper's strategy of first harvesting and second performing optimizations.

**Comment by zhengyangl** [comment]
Will look into this tomorrow.

**Comment by zhengyangl** [comment]
Found the reason for this error.

**Comment by zhengyangl** [comment]
IR before wrong souper pass:
```llvm
define void @_Z8tf_1_foov() local_unnamed_addr #0 {
entry:
  %0 = load i16, i16* @tf_1_var_58, align 2, !tbaa !2
  %conv = zext i16 %0 to i64
  %1 = load i64, i64* @tf_1_var_8, align 8, !tbaa !6
  %or = or i64 %1, %conv
  %tobool = icmp eq i64 %or, 0
  br i1 %tobool, label %entry.if.end_crit_edge, label %if.then

entry.if.end_crit_edge:                           ; preds = %entry
  %.pre = load i64, i64* @tf_1_var_46, align 8, !tbaa !6
  %.pre7 = add nuw nsw i64 %conv, 4294932011
  %.pre8 = and i64 %.pre7, 4294967295
  br label %if.end

if.then:                                          ; preds = %entry
  %2 = load i64, i64* @tf_1_var_46, align 8, !tbaa !6
  %sub = add nuw nsw i64 %conv, 4294932011
  %sh_prom = and i64 %sub, 4294967295
  %shr = lshr i64 %2, %sh_prom
  store i64 %shr, i64* @tf_1_var_168, align 8, !tbaa !6
  br label %if.end

if.end:                                           ; preds = %entry.if.end_crit_edge, %if.then
  %sh_prom4.pre-phi = phi i64 [ %.pre8, %entry.if.end_crit_edge ], [ %sh_prom, %if.then ]
  %3 = phi i64 [ %.pre, %entry.if.end_crit_edge ], [ %2, %if.then ]
  %shr5 = lshr i64 %3, %sh_prom4.pre-phi
  %conv6 = trunc i64 %shr5 to i16
  store i16 %conv6, i16* @tf_1_var_170, align 2, !tbaa !2
  ret void
}
```

**Comment by zhengyangl** [comment]
IR after wrong Souper pass:
```llvm
define void @_Z8tf_1_foov() local_unnamed_addr #0 {
entry:
  %0 = load i16, i16* @tf_1_var_58, align 2, !tbaa !2
  %conv = zext i16 %0 to i64
  %1 = load i64, i64* @tf_1_var_8, align 8, !tbaa !6
  %or = or i64 %1, %conv
  %tobool = icmp eq i64 %or, 0
  br i1 %tobool, label %entry.if.end_crit_edge, label %if.then

entry.if.end_crit_edge:                           ; preds = %entry
  %.pre = load i64, i64* @tf_1_var_46, align 8, !tbaa !6
  %.pre7 = add nuw nsw i64 %conv, 4294932011
  %.pre8 = and i64 4294932011, 4294967295
  br label %if.end

if.then:                                          ; preds = %entry
  %2 = load i64, i64* @tf_1_var_46, align 8, !tbaa !6
  %sub = add nuw nsw i64 %conv, 4294932011
  %sh_prom = and i64 %sub, 4294967295
  %shr = lshr i64 %2, %sh_prom
  store i64 %shr, i64* @tf_1_var_168, align 8, !tbaa !6
  br label %if.end

if.end:                                           ; preds = %entry.if.end_crit_edge, %if.then
  %sh_prom4.pre-phi = phi i64 [ 4294932011, %entry.if.end_crit_edge ], [ %sh_prom, %if.then ]
  %3 = phi i64 [ %.pre, %entry.if.end_crit_edge ], [ %2, %if.then ]
  %shr5 = lshr i64 %3, 4294932011
  %conv6 = trunc i64 %shr5 to i16
  store i16 %conv6, i16* @tf_1_var_170, align 2, !tbaa !2
  ret void
}
```

**Comment by zhengyangl** [comment]
There are three replacement in this pass, 
Replace %.pre8 with 4294932011
Replace %.pre7 with 4294932011
These two replacements are correct. The wrong replacement is 
Replace %sh_prom4.pre-phi with 4294932011

**Comment by zhengyangl** [comment]
This is not a nop replacement.  I can reproduce this error when SOUPER_INFER_NOP is unset.

**Comment by regehr** [comment]
Be careful here! Once a nop is in the cache, Souper will do it even if you don't ask for nop synthesis.

**Comment by regehr** [comment]
Use SOUPER_NO_EXTERNAL_CACHE and see if the bug still happens without SOUPER_INFER_NOP.

**Comment by zhengyangl** [comment]
@regehr You're right, the error disappears. I need some more time to figure out why this replacement happens.

**Comment by zhengyangl** [comment]
Souper generates this optimization.
```%0 = block 2
%1:i16 = var
%2:i64 = zext %1
%3:i64 = addnw 4294932011:i64, %2
%4:i64 = and 4294967295:i64, %3
%5:i64 = phi %0, %4, %4
infer %5
result %4
```

**Comment by regehr** [comment]
This seems ok, right?

**Comment by zhengyangl** [comment]
I believe it missed the path information in the original IR. %conv cannot be a same value in label entry.if.end_crit_edge, and label if.then.

**Comment by zhengyangl** [comment]
The phi node %sh_prom4.pre-phi cannot be optimized here.

**Comment by regehr** [comment]
Are the optimizations individually correct and only wrong when applied together?

**Comment by zhengyangl** [comment]
I guess so, this is a correct optimization, but it is not correct when putting it back.

**Comment by regehr** [comment]
I've been worried about something like this so it's good to have an example. Let's make sure to solve it properly since these problems will only get worse when we synthesize instructions.

**Comment by zhengyangl** [comment]
Sure, I will give a fully reasoned pull request on this error.

**Comment by zhengyangl** [comment]
Phi node handling seems broken (in Solver.cpp). Souper now transform this code:
```llvm
define i32 @func(i32 %x) local_unnamed_addr {
entry:
  %cmp = icmp eq i32 %x, 0
  br i1 %cmp, label %bb1, label %bb2

bb1:                                              ; preds = %entry
  %call_bb1 = add i32 %x, 1
  br label %jmp

bb2:                                              ; preds = %entry
  %call_bb2 = add i32 %x, 1
  br label %jmp

jmp:                                              ; preds = %bb2, %bb1
  %c = phi i32 [ 1, %bb1 ], [ %call_bb2, %bb2 ]
  ret i32 %c
}
```
into 
```llvm
; ModuleID = 'err1.ll'
source_filename = "err1.ll"

define i32 @func(i32 %x) local_unnamed_addr {
entry:
  %cmp = icmp eq i32 %x, 0
  br i1 %cmp, label %bb1, label %bb2

bb1:                                              ; preds = %entry
  %call_bb1 = add i32 %x, 1
  br label %jmp

bb2:                                              ; preds = %entry
  %call_bb2 = add i32 %x, 1
  br label %jmp

jmp:                                              ; preds = %bb2, %bb1
  %c = phi i32 [ 1, %bb1 ], [ %call_bb2, %bb2 ]
  ret i32 1 ; WRONG
}
```

**Comment by zhengyangl** [comment]
Souper first replaces %c with %call_bb1, and then it replaces all uses of %call_bb1 with 1. The second replacement is a wrong replacement.

**Comment by zhengyangl** [comment]
Both replacement seems correctly individually, but I think the first replacement is wrong since the two input [%call_bb1, %bb1] and [%call_bb2, %bb2] are not logically equivalent and they contain path condition information.

**Comment by regehr** [comment]
I think your latest examples above are in the subset of LLVM supported by Alive. Can you try to translate them into Alive and see if it agrees with this optimization?

**Comment by regehr** [comment]
Nevermind, I looked at the example again, the transformation is obviously wrong.

But we should be using Alive whenever possible here. Also, Nuno has a script for automatically converting LLVM into Alive that will be helpful.

**Comment by chenyang78** [comment]
The root cause is that we rely on ```llvm::FoldingSetNodeID``` to compute a unique ID for each souper instruction. So, for the LLVM IR below:

    define i32 @func(i32 %x) local_unnamed_addr {
    entry:
      %cmp = icmp eq i32 %x, 0
      br i1 %cmp, label %bb1, label %bb2

    bb1:                                              ; preds = %entry
      %call_bb1 = add i32 %x, 1
      br label %jmp

    bb2:                                              ; preds = %entry
      %call_bb2 = add i32 %x, 1
      br label %jmp

    jmp:                                              ; preds = %bb2, %bb1
      %c = phi i32 [ %call_bb1, %bb1 ], [ %call_bb2, %bb2 ]
      ret i32 %c
    }


When we build the phi instruction, our ```InstContext::getInst``` routine computes the same ID for ```%call_bb1``` and ```%call_bb2```, because they have the same shape according to how the ID is computed. Consequently, we will get the following Souper IR:

    %0 = block 2
    %1:i32 = var
    %2:i32 = add 1:i32, %1
    %3:i32 = phi %0, %2, %2
    infer %3

Obviously, if souper-infer-nop is *disabled*, we would never infer a value for %3. So we don't see any miscompilation. Unfortunately, when souper-infer-nop is *enabled*, we try to check the following souper IR,

    %0 = block 2
    %1:i32 = var
    %2:i32 = add 1:i32, %1
    %3:i32 = phi %0, %2, %2
    cand %3 %2

and hence get miscompilation.

We might choose not to compute and cache instruction IDs when souper-infer-nop is passed. Or, perhaps just simply remove it.

**Comment by regehr** [comment]
The folding set aspect of souper is something I haven't understood too well. How much benefit does it give us in terms of saving memory and in enabling optimizations? I have no idea.

It would be nice if we could figure out how to disable only problematic folding. As Zhengyang pointed out, the issue stems from the fact that different path information is associated with the two LLVM instructions that got folded into a single souper instruction.

When extracting LLVM into Souper we could do something like refuse to fold instructions that come from different basic blocks, but this is a bit weird since folding should be a property of souper IR, not of LLVM IR. So we would end up needing to add some sort of basic block ID to souper instructions I guess.

**Comment by chenyang78** [comment]
I am not sure about the benefit of folding instructions. It might not be a bad idea that we disable it entirely. If we decide to fold instructions with path information, we need to be really careful. Seems to me if we fold two different LLVM instructions into a single Souper instruction from different paths, we might end up with miscompilation. 

Yes, I think folding should be part of Souper IR (if we decide to keep folding). Otherwise, the chance of introducing mismatches would be high. Expanding Souper IR sounds good to me. On the other hand, the more we add into Souper IR, the more Souper IR would evolve towards LLVM IR. There might be some kind of balance in between, but I don't know where the boundary is.

**Comment by regehr** [comment]
Ok, I just did a small experiment in removing the folding set from InstContext. This basically just works, there are some test case failures but I think they are expected. For example:

```
%0:i1 = var
%1:i1 = var
%2:i1 = and %0, %1
infer %2
```
Here instruction synthesis is supposed to give "result %2" and instead of gives:
```
%3:i1 = and %0, %1
result %3
```
I didn't look at all of the failures but it's at least plausible that they are all like this.

Before we do this for real I want to study the effect on resource usage and optimization power of souper. I'll try to make some time to do this this week.

I'm actually not certain that folding affects optimization power, I had assumed that it did by making it clearer to Souper that two values are actually the same, but that might not be right.

**Comment by regehr** [comment]
I'll add that this would be only the latest in a longish series of changes I've made to Peter's original design to make Souper IR more like LLVM IR, to better support performing complex optimizations in the Pass. For example, I disabled the n-ary operators in favor of binary operators, despite the fact that this restriction will result in some loss of optimization power.

**Comment by regehr** [comment]
If anyone else wants to play, I made this branch:
https://github.com/regehr/souper/tree/no-folding-set
I'll fix up the test cases and do some experiments about resources usage.

**Comment by regehr** [comment]
I'll add that sharing between the LHS and RHS isn't working out for us anyway, due to UB playing a different role on each side. So that particular kind of sharing is really not useful.

**Comment by rsas** [comment]
I'm working on the separation of these two sides among other things. First, we need to separate the KLEE Expr generation from the rest in a clean way. This is tricky as there are many dependencies.

**Comment by regehr** [comment]
@rsas, do you see any bad interactions between your changes and nuking the folding set?

**Comment by regehr** [comment]
Ok, here is an example where turning off folding makes things worse, it's the synthesis result for test/Infer/HackersDelight/P9.opt:

```
; RHS inferred successfully
%4:i32 = ashr %0, 31:i32
%5:i32 = xor %0, %4
%6:i32 = ashr %0, 31:i32
%7:i32 = sub %5, %6
result %7
```
We really don't want to see that repeated ashr. Any ideas for dealing with this? A bad solution would be to add a deduplication pass after synthesis. A better solution would be to avoid synthesizing this.

**Comment by regehr** [comment]
@rsas, is there a constraint in the synthesizer where each instruction is used only once, that we could perhaps turn off, or something similarly easy?

**Comment by regehr** [comment]
P7.opt and P8.opt have the same issue: synthesizing two instructions where one would do

**Comment by regehr** [comment]
All tests pass in my branch now.

**Comment by chenyang78** [comment]
Nice! Thanks for the experiments. John. In terms of the synthesis result below:

```
; RHS inferred successfully
%4:i32 = ashr %0, 31:i32
%5:i32 = xor %0, %4
%6:i32 = ashr %0, 31:i32
%7:i32 = sub %5, %6
result %7
```

It doesn't look like an optimal solution for the given the specification. Seems Souper's synthesizer doesn't take into account the number of a component. For example, for a given component, looks like the synthesizer is allowed to use any number of a valid component. If this is the case, perhaps we would let the synthesizer start from a smaller number, e.g., first try to use one ```ashr``` component, then use two ```ashr``` components, and so on.

Also, I guess we may get an optimal solution by *not* passing -souper-synthesis-ignore-cost, although it might take longer to synthesizing the solution.

**Comment by rsas** [comment]
Hi Yang, synthesis indeed starts with 0 components and increases the number of components as it proceeds. In this example it looks like the reconstruction of the candidate from the model references two different instructions somehow leading to this result. I rely on the folded set ID to reconstruct unique candidates and also to detect seen candidates for synthesis bug detection. For constraints it does not matter, we can have duplicates there. I’ll look into this soon in more detail.

Raimondas

**Comment by chenyang78** [comment]
@rsas I see. Thanks for the explanation.

**Comment by regehr** [comment]
It seems clear that Souper is going to have to be more careful in how it treats phi nodes: it just isn't the case that two things that look alike must be the same. Ideas appreciated...

**Comment by chenyang78** [comment]
Argh, thanks for the example. Looks like we have to get back to the original proposal - what we need is perhaps some kind of information (path information or block information or something else) that could be used to separate these two optimizations, i.e. avoiding the interference of the first optimization on the second one. Note that we can't simply invalidate the second optimization, because it obviously gives us more benefits 
than performing the first one.

**Comment by rsas** [comment]
What about adding block predicates directly to the Phi inst instead to the Block?

Raimondas

**Comment by regehr** [comment]
Can you sketch this out Raimondas? I'm having trouble picturing it.

**Comment by rsas** [comment]
I have to see how the ID is calculated and what additional information we could add to make it unique. I thought about predicates controlling the path, not sure it would work.

Raimondas

**Comment by regehr** [comment]
well, it would be good to find the right solution but for the moment I propose this fix:
https://github.com/google/souper/pull/306

**Comment by chenyang78** [comment]
@regehr Sounds good to me. Thanks.
